File: experimental/mlf-type-system/src/mlf/prefix.ts
import {BindingMap} from '../utils/bindings';
import {Err, Ok, Result} from '../utils/result';

import {Bound, Polytype, Type} from './type';

export type PrefixError =
  | {
      readonly kind: 'UnboundTypeVariable';
      readonly name: string;
    }
  | {
      readonly kind: 'InfiniteType';
    }
  | {
      readonly kind: 'InvalidTypeUpdate';
      readonly old: Type;
      readonly new: Type;
    };

/**
 * State required to type-check a Brite application. Holds things like allocated
 * type variables.
 */
export class Prefix {
  // The id for the type variable in our program.
  private nextType = 0;

  // All the allocated type variables in our program.
  private readonly typeVariables = new Map<
    string,
    {level: number; bound: Bound}
  >();

  // A stack of all our levels and the types allocated in each level.
  private readonly levels: Array<Set<string>> = [];

  /**
   * Gets the current level for our program. Levels are used to determine which
   * type variables need to be generalized.
   *
   * Think of it a bit like garbage collection. We allocate a type variable at a
   * given level. Then when we exit that level all type variables need to be
   * “collected” or in our case “quantified”.
   */
  getLevel(): number {
    return this.levels.length;
  }

  /**
   * Increments the level.
   */
  incrementLevel() {
    this.levels.push(new Set());
  }

  /**
   * Decrements the level. Removes all variables allocated in this level from
   * our prefix. Make sure you don’t need any of the variables before they
   * are removed!
   */
  decrementLevel() {
    const variables = this.levels.pop();
    if (variables !== undefined) {
      for (const variable of variables) {
        this.typeVariables.delete(variable);
      }
    }
  }

  /**
   * Are there no type variables in our prefix?
   */
  isEmpty(): boolean {
    return this.typeVariables.size === 0;
  }

  /**
   * Allocates a fresh type variable in our prefix with no bound. When
   * `Prefix.decrementLevel()` is called this type will be deallocated.
   *
   * The created type is guaranteed to have a unique name in our entire program
   * since the generated name starts with `$` which is not allowed in
   * user identifiers.
   */
  fresh(): string {
    return this.freshWithBound(Type.unbounded);
  }

  /**
   * Allocates a type variable in our prefix with the provided bound. When
   * `Prefix.decrementLevel()` is called this type will be deallocated.
   *
   * The created type is guaranteed to have a unique name in our entire program
   * since the generated name starts with `$` which is not allowed in
   * user identifiers.
   */
  freshWithBound(bound: Bound): string {
    const name = `$${this.nextType++}`;
    this.addToLevel(name, bound);
    return name;
  }

  /**
   * Adds a named type variable to the prefix. If a type variable with this name
   * already exists in the prefix then a new name is generated. The name used to
   * allocate the type variable is returned. When `Prefix.decrementLevel()` is
   * called this type variable will be deallocated.
   */
  add(name: string, bound: Bound): string {
    if (this.typeVariables.has(name)) {
      return this.freshWithBound(bound);
    } else {
      this.addToLevel(name, bound);
      return name;
    }
  }

  /**
   * Internal function for adding a type variable to our prefix.
   */
  private addToLevel(name: string, bound: Bound): void {
    this.typeVariables.set(name, {level: this.levels.length, bound});
    if (this.levels.length > 0) this.levels[this.levels.length - 1].add(name);
  }

  /**
   * Looks up a type variable in our prefix. The type variable must have come
   * from `Prefix.fresh()` or `Prefix.freshWithBound()` as to assure no
   * scoping shenanigans are necessary since all type names are guaranteed to be
   * unique in the program.
   */
  lookup(
    name: string
  ): Result<{readonly level: number; readonly bound: Bound}, PrefixError> {
    const entry = this.typeVariables.get(name);
    if (entry === undefined) {
      return Err<PrefixError>({kind: 'UnboundTypeVariable', name});
    } else {
      return Ok({level: entry.level, bound: entry.bound});
    }
  }

  /**
   * Updates a type variable in our prefix. Assumes that the old type and the
   * new provided type are in the instance relation (`(Q) old ⊑ new`). Please
   * don’t call this function outside of the unification algorithm. Since the
   * unification algorithm takes care to maintain the relation `old ⊑ new`.
   *
   * Possibly returns an error if the occurs check or abstraction check fails.
   *
   * Also, if a free type variable in the provided bound has a higher level then
   * the type variable we are updating then we “level up” the type variable in
   * the bound so that it won’t be deallocated before the type variable we are
   * updating.
   */
  update(name: string, bound: Bound): PrefixError | undefined {
    const entry = this.typeVariables.get(name);
    if (entry === undefined) {
      throw new Error(`Type variable not found: "${name}"`);
    }
    // Checks to see if the type variable we are updating exists in the type we
    // are updating to. If so then we have an infinite type. Fail early!
    if (this.occursCheck(name, bound.type)) return {kind: 'InfiniteType'};
    // TODO: Documentation comment
    if (
      entry.bound.kind === 'rigid' &&
      !this.abstractionCheck(entry.bound.type, bound.type)
    ) {
      return {
        kind: 'InvalidTypeUpdate',
        old: entry.bound.type,
        new: bound.type,
      };
    }
    // Update the entry’s bound to the new bound.
    entry.bound = bound;
    // Level up all the free type variables in our bound’s type.
    this.levelUp(entry.level, bound.type);
    return undefined;
  }

  /**
   * Moves the free variables of the provided type at a level higher then the
   * provided level to the provided level. Also updates the bounds of these free
   * variables recursively.
   *
   * This operation saves type variables that are needed at the provided level
   * from deallocation when `Prefix.decrementLevel()` is called.
   *
   * And yes, pun intended.
   */
  private levelUp(newLevel: number, type: Polytype) {
    // For all the free variables in our type...
    for (const name of Type.getFreeVariables(type)) {
      const entry = this.typeVariables.get(name);
      if (entry === undefined) continue;
      // If the entry’s level is larger than the new level then that means this
      // type variable will be deallocated before we need it at `newLevel`. So
      // we need to update its level.
      if (entry.level > newLevel) {
        // Update the level.
        const oldLevel = entry.level;
        entry.level = newLevel;
        // Move the type variable from its old level to its new level so it
        // won’t be deallocated with the old level.
        if (oldLevel > 0) this.levels[oldLevel - 1].delete(name);
        if (newLevel > 0) this.levels[newLevel - 1].add(name);
        // Recursively update the free type variables in our bound since they
        // too might be deallocated too soon if we don’t.
        this.levelUp(newLevel, entry.bound.type);
      }
    }
  }

  /**
   * Recursively tests to see if the provided type or any of the type’s bounds
   * in prefix contain the provided name. If so then it is not safe to update
   * the provided name with the provided type.
   */
  private occursCheck(testName: string, type: Polytype): boolean {
    // For all free type variables...
    for (const name of Type.getFreeVariables(type)) {
      // If this name is the one we are testing for then the test name does
      // occur in our type! Uh oh...
      if (name === testName) return true;
      // Get the bound for this type variable.
      const entry = this.typeVariables.get(name);
      if (entry === undefined) continue;
      // Recursively check to see if our name occurs in the type bound.
      if (this.occursCheck(testName, entry.bound.type)) return true;
    }
    return false;
  }

  private abstractionCheck(oldType: Polytype, newType: Polytype): boolean {
    const prefix = this;
    const scope1 = new BindingMap<string, Polytype>();
    const scope2 = new BindingMap<string, Polytype>();

    if (!projectionEquals(oldType, newType)) return false;
    if (Type.isMonotype(oldType)) return true;

    return true;
    // if (newType.description.kind === 'Variable') {
    // } else {
    //   return true;
    // }

    /**
     * Counts the number of bottom types which appear in a flexible context.
     *
     * For illustration consider the following type. For readability we write
     * the type on multiple lines:
     *
     * ```
     * ∀(a ≥ ⊥).
     * ∀(b ≥ ∀(b1 ≥ ⊥, b2 = ⊥).b1 → b2).
     * ∀(c = ∀(c1 ≥ ⊥, c2 = ⊥).c1 → c2).
     * a → b → c
     * ```
     *
     * In this type we have five bottom types (⊥). Each bottom type has a “path”
     * corresponding to:
     *
     * 1. `a`: ≥
     * 2. `b1`: ≥≥
     * 3. `b2`: ≥=
     * 4. `c1`: =≥
     * 5. `c2`: ==
     *
     * We consider a bottom type to appear in a flexible context if the path
     * starts with one or more flexible bounds (`≥`). So `a`, `b1`, and `b2`
     * appear in a flexible context since they begin with one or more
     * flexible bounds.
     *
     * For more information about the theory of “flexible paths” see
     * [MLF thesis][1] section 2.7.2 entitled “Weights”. For the theory behind
     * paths. See definition 2.7.3 and its accompanying definition of w(t) for
     * an application of weights. See figure 4.1 and lemma 2.7.8 for how the
     * theory of weights applies to our specific abstraction check algorithm.
     *
     * [1]: https://pastel.archives-ouvertes.fr/file/index/docid/47191/filename/tel-00007132.pdf
     */
    function flexiblePaths(type: Polytype): number {
      let paths = 0;

      while (type.description.kind === 'Quantify') {
        if (type.description.bound.kind === 'flexible') {
          paths += allPaths(type.description.bound.type);
        }
        type = type.description.body;
      }

      return paths;

      function allPaths(type: Polytype): number {
        let paths = 0;

        while (type.description.kind === 'Quantify') {
          paths += allPaths(type.description.bound.type);
          type = type.description.body;
        }

        if (type.description.kind === 'Bottom') {
          paths += 1;
        }

        return paths;
      }
    }

    /**
     * Determine if the projection of two types under our prefix are equal. The
     * projection is a simple inlining of all bounds no matter their kind.
     */
    function projectionEquals(t1: Polytype, t2: Polytype): boolean {
      // Substitute the type for `t1` if it is in scope.
      if (t1.description.kind === 'Variable') {
        const type = scope1.get(t1.description.name);
        if (type !== undefined) return projectionEquals(type, t2);
      }

      // Substitute the type for `t2` if it is in scope.
      if (t2.description.kind === 'Variable') {
        const type = scope2.get(t2.description.name);
        if (type !== undefined) return projectionEquals(t1, type);
      }

      // If both `t1` and `t2` are variables and they share the same name then
      // we know their bounds to be equal.
      if (
        t1.description.kind === 'Variable' &&
        t2.description.kind === 'Variable' &&
        t1.description.name === t2.description.name
      ) {
        return true;
      }

      // Substitute the type for `t1` from our prefix. Return false if the type
      // is not found in our prefix.
      if (t1.description.kind === 'Variable') {
        const entry = prefix.typeVariables.get(t1.description.name);
        if (entry === undefined) return false;
        return projectionEquals(entry.bound.type, t2);
      }

      // Substitute the type for `t2` from our prefix. Return false if the type
      // is not found in our prefix.
      if (t2.description.kind === 'Variable') {
        const entry = prefix.typeVariables.get(t2.description.name);
        if (entry === undefined) return false;
        return projectionEquals(t1, entry.bound.type);
      }

      // Unwrap `t1` quantifications into scope.
      if (t1.description.kind === 'Quantify') {
        let pops = 0;
        while (t1.description.kind === 'Quantify') {
          pops++;
          scope1.push(t1.description.name, t1.description.bound.type);
          t1 = t1.description.body;
        }
        const equals = projectionEquals(t1, t2);
        for (let i = 0; i < pops; i++) scope1.pop();
        return equals;
      }

      // Unwrap `t2` quantifications into scope.
      if (t2.description.kind === 'Quantify') {
        let pops = 0;
        while (t2.description.kind === 'Quantify') {
          pops++;
          scope2.push(t2.description.name, t2.description.bound.type);
          t2 = t2.description.body;
        }
        const equals = projectionEquals(t1, t2);
        for (let i = 0; i < pops; i++) scope2.pop();
        return equals;
      }

      // Check equality of scalar types.
      if (t1.description.kind === 'Bottom') return t2.description.kind === 'Bottom'; // prettier-ignore
      if (t2.description.kind === 'Bottom') return false;
      if (t1.description.kind === 'Boolean') return t2.description.kind === 'Boolean'; // prettier-ignore
      if (t2.description.kind === 'Boolean') return false;
      if (t1.description.kind === 'Number') return t2.description.kind === 'Number'; // prettier-ignore
      if (t2.description.kind === 'Number') return false;
      if (t1.description.kind === 'String') return t2.description.kind === 'String'; // prettier-ignore
      if (t2.description.kind === 'String') return false;

      // Check equality of composite types.
      if (t1.description.kind === 'Function') {
        if (t2.description.kind === 'Function') {
          return (
            projectionEquals(t1.description.parameter, t2.description.parameter) && // prettier-ignore
            projectionEquals(t1.description.body, t2.description.body)
          );
        } else {
          return false;
        }
      } else if (t2.description.kind === 'Function') {
        return false;
      }

      const never1: never = t1.description;
      const never2: never = t2.description;
      return never1 && never2;
    }
  }

  /**
   * Prints the prefix to a display string using the standard syntax for types
   * in academic literature. Particularly the [MLF][1] paper we implement.
   *
   * Brite programmers will not be familiar with this syntax. It is for
   * debugging purposes only.
   *
   * [1]: http://pauillac.inria.fr/~remy/work/mlf/icfp.pdf
   */
  toDisplayString() {
    if (this.typeVariables.size === 0) return '(∅)';
    // All the type variables in our prefix in dependency order.
    const seen = new Set<string>();
    // Visits a type variable with the provided name. Adds the dependencies of
    // the visited type variable before adding the type variable name itself.
    const visit = (name: string) => {
      if (seen.has(name)) return;
      const entry = this.typeVariables.get(name);
      if (entry === undefined) return;
      for (const name of Type.getFreeVariables(entry.bound.type)) visit(name);
      seen.add(name);
    };
    // Visit all the type variables in our prefix. This will populate the `seen`
    // set in dependency order. That is dependencies will appear before their
    // dependents in the set.
    for (const name of this.typeVariables.keys()) visit(name);
    // Iterate through the type variables in dependency order and produce the
    // display string.
    const bounds: Array<string> = [];
    for (const name of seen) {
      const entry = this.typeVariables.get(name);
      if (entry === undefined) continue;
      const bound = entry.bound;
      if (bound.kind === 'flexible' && bound.type === undefined) {
        bounds.push(name);
      } else {
        const kind = bound.kind === 'flexible' ? '≥' : '=';
        const type =
          bound.type !== undefined ? Type.toDisplayString(bound.type) : '⊥';
        bounds.push(`${name} ${kind} ${type}`);
      }
    }
    // Return the prefix in parentheses.
    return `(${bounds.join(', ')})`;
  }
}

File: src/Checker/Prefix.ml
type level = {
  index: int;
  names: (string, unit) Hashtbl.t;
}

type entry = {
  mutable level: level;
  mutable bound: Type.bound;
}

type t = {
  mutable counter: int;
  mutable levels: level list;
  entries: (string, entry) Hashtbl.t;
}

(* Creates a new prefix. *)
let create () =
  {
    counter = 1;
    levels = [];
    entries = Hashtbl.create 1000;
  }

(* Introduces a new level for the execution of the provided function. Cleans up
 * the level after the function finishes. Any type variables created in the
 * level will be removed from the prefix unless they were re-ordered in
 * an update. *)
let level prefix f =
  (* Add a new level to the prefix. *)
  let index = match prefix.levels with [] -> 0 | level :: _ -> level.index + 1 in
  prefix.levels <- { index; names = Hashtbl.create 16 } :: prefix.levels;
  let counter = prefix.counter in
  (* Execute our function with the new level. *)
  let result = try Ok (f ()) with e -> Error e in
  (* Remove the new level from the prefix. *)
  let level = List.hd prefix.levels in
  prefix.levels <- List.tl prefix.levels;
  prefix.counter <- counter;
  (* Delete all type variables at this level from the prefix. *)
  Hashtbl.iter (fun name () -> Hashtbl.remove prefix.entries name) level.names;
  (* If this was the last level then make sure there are no more entries in
   * the prefix. *)
  if prefix.levels = [] && not (Hashtbl.length prefix.entries = 0) then (
    failwith "Expected all type variables to be cleaned up."
  );
  (* Return the result from executing our function. *)
  match result with
  | Ok result -> result
  | Error error -> raise error

(* Private function for adding a type variable to the current level. If there is
 * no current level then we panic. *)
let add_to_level prefix name bound =
  (* Panic if we are not in a level or if a type variable with the same name
   * already exists in the prefix. *)
  if prefix.levels = [] then failwith "Type variable must be added to the prefix in a level.";
  if Hashtbl.mem prefix.entries name then failwith "Type variable with the same name already exists in the prefix.";
  (* Check to make sure that every free type variable has a corresponding entry
   * in the prefix. *)
  StringSet.iter
    (fun name -> ignore (Hashtbl.find prefix.entries name))
    (Lazy.force bound.Type.bound_type.polytype_free_variables);
  (* Add the type variable to the current level. *)
  let level = List.hd prefix.levels in
  Hashtbl.add level.names name ();
  (* Add the type variable to the prefix. *)
  Hashtbl.add prefix.entries name { level; bound }

(* IMPORTANT: It is expected that all free type variables are bound in the
 * prefix! If this is not true then we will panic.
 *
 * Creates a type variable with a unique name in the prefix and the provided
 * bound. Returns the new unique name. *)
let fresh_with_bound prefix bound =
  (* Directly return monotypes since they are always inlined in normal form. *)
  match bound.Type.bound_type.polytype_description with
  | Monotype bound_type -> bound_type
  | _ ->
    let rec unique () =
      let name = "t" ^ string_of_int prefix.counter in
      prefix.counter <- prefix.counter + 1;
      if Hashtbl.mem prefix.entries name then unique () else name
    in
    let name = unique () in
    add_to_level prefix name bound;
    Type.variable_with_kind name (Type.kind bound.bound_type)

(* Creates a type variable with a unique name in the prefix with a bottom bound.
 * Returns the new unique name. *)
let fresh prefix = fresh_with_bound prefix Type.unbounded

(* Adds a bound to the prefix with the provided name. If the name already exists
 * in the prefix then generate a new name so that all type variables in the
 * prefix are unique. Return the new name if we had to generate it. *)
let add prefix name bound =
  let name' = Namer.unique (Hashtbl.mem prefix.entries) name in
  add_to_level prefix name' bound;
  if name = name' then None else Some (Type.variable_with_kind name' (Type.kind bound.bound_type))

(* Finds the bound for the provided name in the prefix. If no type variable
 * could be found then we panic. The bound returned will always be in
 * normal form. *)
let lookup prefix name =
  let entry = Hashtbl.find prefix.entries name in
  if not entry.bound.bound_type.polytype_normal then (
    match Type.normal entry.bound.bound_type with
    | None -> ()
    | Some t -> entry.bound <- Type.bound entry.bound.bound_flexibility t
  );
  entry.bound

(* Merges the bounds of a quantified type into the prefix. If any of the bound
 * names already exist in the prefix then we need to generate new names. Those
 * names will be substituted in the bounds and the provided body type. The new
 * body type with substituted names will be returned. *)
let instantiate prefix bounds body =
  let substitutions = Nel.fold_left (fun substitutions (name, bound) -> (
    (* If we have some substitutions then apply them to our bound. *)
    let bound = match Type.substitute_polytype substitutions bound.Type.bound_type with
    | None -> bound
    | Some bound_type -> Type.bound bound.bound_flexibility bound_type
    in
    (* Add the bound to our prefix. If we had to generate a new name then add a
     * substitution to our map.
     *
     * It is worth considering how we handle the case
     * `∀a.∀(a = a).∀(a = a). ...`. In this type the second binding for `a`
     * will be generated a unique name and we will add a substitution for that
     * name. In the third binding for `a` we will also generate a unique name
     * and we will override the previous substitution. The issue we want to
     * avoid here is not updating the substitutions map for the third `a`. *)
    let new_type = add prefix name bound in
    let substitutions = match new_type with
    | Some new_type -> StringMap.add name new_type substitutions
    | None -> substitutions
    in
    substitutions
  )) StringMap.empty bounds in
  (* Substitute the body type and return if we have some substitutions. *)
  match Type.substitute_monotype substitutions body with Some t -> t | None -> body

(* Generalizes a type at the current level of the prefix. If the type references
 * variables in the prefix at the current level then we will quantify the
 * provided monotype with those type variables. If the type references variables
 * at another level then we will not quantify those. The returned type will
 * be in normal form. *)
let generalize prefix t =
  (* Get the current prefix level. Panic if we are not in a level. *)
  if prefix.levels = [] then failwith "May only generalize in a level.";
  let level = List.hd prefix.levels in
  (* Create a hash table for tracking the type variables we’ve visited. Also
   * create a list of the bounds we are going to quantify. *)
  let visited = Hashtbl.create 10 in
  let bounds = ref [] in
  (* Our visitor will look at each free type variable and possibly add it to our
   * bounds list and recurse. *)
  let rec visit name =
    (* If we have not yet visited this type variable... *)
    if not (Hashtbl.mem visited name) then (
      Hashtbl.add visited name ();
      (* If the type variable is on our current level... *)
      let entry = Hashtbl.find prefix.entries name in
      if entry.level.index >= level.index then (
        (* Recurse and visit all free variables in our bound because the bound
         * might have some dependencies on this level. *)
        StringSet.iter visit (Lazy.force entry.bound.bound_type.polytype_free_variables);
        (* Add our bound to the list we will use to quantify. It is important
         * we do this after recursing! Since the bound we add depends on the
         * bounds we might add while recursing. *)
        bounds := (name, entry.bound) :: !bounds
      )
    )
  in
  (* Visit all the free variables in this type. *)
  StringSet.iter visit (Lazy.force t.Type.monotype_free_variables);
  (* Quantify the type by the list of bounds we collected. Also convert the type
   * to normal form. *)
  let t = if !bounds = [] then Type.to_polytype t else Type.quantify (Nel.from_list (List.rev !bounds)) t in
  let t = match Type.normal t with Some t -> t | None -> t in
  t

exception Occurs

(* Checks to see if the provided name occurs anywhere in the type or in the
 * bounds of any free type variables recursively. If it does then we can’t
 * update the type variable at `name` with the provided type.
 *
 * NOTE: This function may get a bit hot. Could be a good function to make
 * incremental and improve performance. *)
let occurs prefix name t =
  (* Don’t allocate a hash table if there are no free variables in `t`. *)
  if StringSet.is_empty (Lazy.force t.Type.polytype_free_variables) then (
    false
  ) else (
    (* Cache for type variables we have seen before so we don’t need to check
     * them twice. Without this cache the occurs algorithm is worst case O(n²)
     * instead of O(n). Consider `∀(a, b = a → a).b → b`. Without caching the
     * fact that we’ve seen `b` before we might visit `a` four times. Two times
     * each time we visit `b`. *)
    let seen = Hashtbl.create 10 in
    let rec loop t =
      Lazy.force t.Type.polytype_free_variables |> StringSet.iter (fun name' -> (
        if String.equal name name' then (
          raise Occurs
        ) else if not (Hashtbl.mem seen name') then (
          Hashtbl.add seen name' ();
          let entry = Hashtbl.find prefix.entries name' in
          loop (entry.bound.Type.bound_type)
        )
      ))
    in
    (* Break out of our loop with an exception. *)
    try loop t; false with Occurs -> true
  )

(* “Re-orders” a type and its dependencies in the prefix. The prefix is ordered
 * by levels. All the type variable in a given level will be generalized
 * together. So if our new level comes before the level of our type’s
 * dependencies then we need to move our type’s dependencies to the new level.
 * Otherwise we’ll have dangling pointers when the type’s dependencies’ level
 * is removed from our prefix.
 *
 * Oh, and yes, pun intended. *)
let rec level_up prefix new_level t =
  Lazy.force t.Type.polytype_free_variables |> StringSet.iter (fun name -> (
    let entry = Hashtbl.find prefix.entries name in
    let old_level = entry.level in
    (* We know our old level will be removed before our new level if the index
     * is larger. *)
    if old_level.index > new_level.index then (
      (* Change our type variable’s level. *)
      entry.level <- new_level;
      Hashtbl.remove old_level.names name;
      Hashtbl.add new_level.names name ();
      (* Recurse because we might have dependencies in our bound that also need
       * to be updated. *)
      level_up prefix new_level (entry.bound.Type.bound_type)
    )
  ))

(* Performs the checks which must pass for an update to be safe. *)
let update_check prefix name old_bound new_bound =
  if (
    (* Run an “occurs” check to make sure that we aren’t creating an infinite
     * type with this update. If we would create an infinite type then return
     * an error. *)
    occurs prefix name new_bound.Type.bound_type
  ) then (
    Error (Diagnostics.report_error (InfiniteType { name; type_ = new_bound.bound_type }))
  ) else if (
    (* If the old bound is rigid then we check to make sure that the new type
     * is an abstraction of the old type. If it is not then we have an
     * invalid update!
     *
     * NOTE: We re-use the `IncompatibleTypes` error here, but maybe we want a
     * more descriptive error? *)
    old_bound.Type.bound_flexibility == Rigid &&
    not (Abstraction.check (lookup prefix) old_bound.bound_type new_bound.bound_type)
  ) then (
    let type1 = old_bound.bound_type in
    let type2 = new_bound.bound_type in
    Error (Diagnostics.report_error (IncompatibleTypes { type1; type2 }))
  ) else (
    (* The update is ok. You may proceed to commit changes... *)
    Ok ()
  )

(* IMPORTANT: We assume that `(Q) t1 ⊑ t2` holds. Do not call this function with
 * two types which do not uphold this relation!
 *
 * The update algorithm replaces the bound of a type variable in our prefix with
 * a new one. This operation only succeeds if the prefix after updating is
 * strictly an instance of the prefix before updating. In other words `Q ⊑ Q'`
 * must hold. We also must not add a circular dependency to our prefix as this
 * breaks the theoretical prefix ordering. So our prefix has these invariants:
 *
 * 1. If `name` does not already exist in the prefix we panic.
 * 2. If `bound` or any of the referenced type variables in `bound` contain a
 *    reference to `name` we return an error.
 * 3. If `Q ⊑ Q'` would not hold after applying this update we return an error.
 *
 * We assume that the new type is an instance of the old type. However, we do
 * check to ensure that the new type is an abstraction of the old type. *)
let update prefix name bound =
  let entry = Hashtbl.find prefix.entries name in
  (* Make sure our update is safe. *)
  match update_check prefix name entry.bound bound with
  (* If the update is not safe then return an error without
   * committing anything. *)
  | Error error -> Error error
  (* If the update is safe then update our entry and update the levels of our
   * new type’s dependencies. *)
  | Ok () ->
    entry.bound <- bound;
    level_up prefix entry.level bound.bound_type;
    Ok ()

(* IMPORTANT: We assume that `(Q) t1 ⊑ t3` and `(Q) t2 ⊑ t3` hold. Do not call
 * this function with types which do not uphold this relation!
 *
 * Updates two types to the same bound. Corresponds to the merge algorithm in
 * the [MLF Thesis][1] along with two calls to update. It is important we do
 * these two updates together in a transaction so a single error means we don’t
 * commit anything. *)
let update2 prefix name1 name2 bound =
  let entry1 = Hashtbl.find prefix.entries name1 in
  let entry2 = Hashtbl.find prefix.entries name2 in
  (* Make sure both our updates are safe. *)
  match (
    update_check prefix name1 entry1.bound bound,
    update_check prefix name2 entry2.bound bound
  ) with
  (* If the update is not safe then return an error without
   * committing anything. *)
  | Error error, _ -> Error error
  | _, Error error -> Error error
  (* If the update is safe then update our entry and update the levels of our
   * new type’s dependencies. *)
  | Ok (), Ok () ->
    (* Update one of the entries to equal the provided bound. Then, link the
     * other entry to updated entry. We let the entry which will live longer
     * be the one to hold our bound. We let the entry which will live shorter
     * be our link. *)
    if entry1.level.index <= entry2.level.index then (
      entry1.bound <- bound;
      entry2.bound <- Type.bound Type.Rigid (Type.to_polytype (Type.variable_with_kind name1 (Type.kind bound.bound_type)));
      level_up prefix entry1.level bound.bound_type;
      Ok ()
    ) else (
      entry2.bound <- bound;
      entry1.bound <- Type.bound Type.Rigid (Type.to_polytype (Type.variable_with_kind name2 (Type.kind bound.bound_type)));
      level_up prefix entry2.level bound.bound_type;
      Ok ()
    )

(* Collects all the current bounds of the prefix into the list. The bounds are
 * sorted in dependency order. That is, dependents are listed after
 * their dependencies. *)
let bounds prefix =
  let visited: (string, unit) Hashtbl.t = Hashtbl.create (Hashtbl.length prefix.entries) in
  let rec visit (bounds: (string * Type.bound) list) name =
    if Hashtbl.mem visited name then (
      bounds
    ) else (
      Hashtbl.add visited name ();
      let { bound; _ } = Hashtbl.find prefix.entries name in
      let bounds = StringSet.fold
        (fun name bounds -> visit bounds name)
        (Lazy.force bound.bound_type.polytype_free_variables)
        bounds
      in
      let bounds = (name, bound) :: bounds in
      bounds
    )
  in
  let bounds = Hashtbl.fold (fun name _ bounds -> visit bounds name) prefix.entries [] in
  List.rev bounds
